File: java/two_sum.java
import java.util.Arrays;
import java.util.Comparator;

public class HelloWorld {
	class Pair {
		int val;
		int index;
	}
	public int[] twoSum(int[] nums, int target) {              
       Pair[] pairs=new Pair[nums.length];
	   for (int i=0; i<nums.length; i++) {
		   pairs[i]=new Pair();
    	   pairs[i].val=nums[i];
    	   pairs[i].index=i;
       }
	   
	   Arrays.sort(pairs, new Comparator<Pair>(){
		   public int compare(Pair p1, Pair p2) {
			   return p1.val-p2.val;
		   }
	   });
	   
       int l=0, r=nums.length-1;
       while (l<r) {
    	   if (pairs[l].val+pairs[r].val==target) return new int [] {pairs[l].index, pairs[r].index};
    	   else if (pairs[l].val+pairs[r].val<target) l++;
    	   else r--;
       }
       
       return new int [] {0, 0};
    }
	
	public static void main(String[] args) {
		int [] nums={3,2,4};
		HelloWorld h = new HelloWorld();
		int [] res = h.twoSum(nums, 6);
		System.out.println(Arrays.toString(res));
	}
}

2. HashMap
public class Solution {
	public int[] twoSum(int[] nums, int target) {              
        Map<Integer, Integer> map=new HashMap<>();
        int[] res={0,0};
        for (int i=0; i<nums.length; i++) {
            map.put(nums[i], i);
        }
        
        for (int i=0; i<nums.length; i++) {
            if (map.containsKey(target-nums[i]) && map.get(target-nums[i])!=i) {
                res[0]=i;
                res[1]=map.get(target-nums[i]);
                break;
            }
        }
        
        return res;
    }
}

File: golang/two-sum.go
1. hash table
func twoSum(nums []int, target int) []int {
	res := make([]int, 2)
	table := map[int]int{}

	for i, num := range nums {
		if index, ok := table[target-num]; ok {
			res[0] = index
			res[1] = i
			return res
		}
		table[num] = i
	}
	return res
}

2. two pointers
type Element struct {
	index int
	val   int
}
type ByVal []Element

func (a ByVal) Len() int           { return len(a) }
func (a ByVal) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByVal) Less(i, j int) bool { return a[i].val < a[j].val }

func twoSum(nums []int, target int) []int {
	res := make([]int, 2)
	elements := make([]Element, len(nums))

	for i, num := range nums {
		elements[i].index = i
		elements[i].val = num
	}
	sort.Sort(ByVal(elements))

	for l, r := 0, len(nums)-1; l < r; {
		if elements[l].val+elements[r].val == target {
			res[0] = elements[l].index
			res[1] = elements[r].index
			return res
		} else if elements[l].val+elements[r].val < target {
			l = l + 1
		} else {
			r = r - 1
		}
	}

	return res
}

